### PR TITLE
Enable trace by OCL_ICD_ENABLE_TRACE environment variable

### PR DESCRIPTION
**Pull request by cdai2**

### CODE REVIEW

**Comment by Kerilk** [comment]
@jenatali Would you be able to take a look?
This impacts Windows too (see https://github.com/KhronosGroup/OpenCL-ICD-Loader/issues/167).

File: loader/icd.c
Comment (Kerilk) [inline loader/icd.c:29]: I would use an integer here to store the state of the tracing activation.
The string returned by getenv should not change through the execution, so better compare it early.

File: loader/icd.c
Comment (Kerilk) [inline loader/icd.c:38]: Two remarks here

- Same remark as in the issue, discussion of using `khrIcd_getenv` vs `khrIcd_secure_getenv`. Tagging @jenatali as the windows implications seem stronger than the Linux ones (I may be wrong).
- the initialization of the tracing variable should go in it's own subroutine, maybe per os, depending on the result of the above discussion. This initialization should end up with `khrEnableTrace` containing a boolean value signifying if the tracing is active or not. The list of acceptable values for the string will need to be determined but will most probably end being more than just `"True"`

File: loader/icd.h
Comment (Kerilk) [inline loader/icd.h:87]: see above comment regarding variable type.

File: loader/icd.h
Comment (Kerilk) [inline loader/icd.h:165]: I would rather see something like this:
```C
if (khrEnableTrace) \
    do \
    {
        fprintf(stderr, "KHR ICD trace at %s:%d: ", __FILE__, __LINE__); \
        fprintf(stderr, __VA_ARGS__); \
    } while (0)
```
As this will keep the established usage of requiring a `;` after each macro invocation. I also simplified the test to reflect the usage of the boolean/int variable.

File: loader/icd.h
Comment (Kerilk) [inline loader/icd.h:191]: Same as above.

File: loader/icd.h
Comment (Kerilk) [inline loader/icd.h:203]: Same as above.

File: loader/icd.h
Comment (Kerilk) [inline loader/icd.h:220]: Same as above.

File: loader/linux/icd_linux.c
Comment (Kerilk) [inline loader/linux/icd_linux.c:175]:
```suggestion
        KHR_ICD_TRACE("Failed to load driver because %s.\n", dlerror());
```

**Review by Kerilk — CHANGES_REQUESTED** [review]
Thanks for the PR, it is greatly appreciated.
I do have some comments/change requests, and also requested others review those changes.

File: loader/linux/icd_linux.c
Comment (cdai2) [reply loader/linux/icd_linux.c]: Fixed.

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: Fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: Fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: Fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: Fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: Fixed

File: loader/icd.c
Comment (cdai2) [reply loader/icd.c]: Fixed. I defined it as integer.

File: loader/icd.c
Comment (cdai2) [reply loader/icd.c]: For item 1,  I will wait for @jenatali 's feedback.  Personally I still think this env variable does not change the execution logic of loading drivers. So we do not need the khrIcd_secure_getenv.

For item 2, I added "true", "T", "1".   I added the initialization of trace enabling because khrIcdInitialize is the entry point to initialize ICD.

**Comment by cdai2** [comment]
@Kerilk Thanks for your review.  I fixed most of issues according to your comments.

File: loader/icd_dispatch.c
Comment (Kerilk) [inline loader/icd_dispatch.c:20]: This file should be reverted.
```suggestion
```

File: loader/icd_dispatch_generated.c
Comment (Kerilk) [inline loader/icd_dispatch_generated.c:25]: This file should be reverted.
```suggestion
#include "icd_dispatch.h"
#include "icd.h"

#ifdef __cplusplus
```

File: loader/windows/icd_windows_apppackage.cpp
Comment (Kerilk) [inline loader/windows/icd_windows_apppackage.cpp:25]: Not sure about this one (don't know why it would be required), but it would seem more coherent with below conventions if it were written as:

```suggestion
extern "C" int khrEnableTrace;

```

File: loader/icd.c
Comment (Kerilk) [inline loader/icd.c:46]: This should go in its own subroutine. also, line 44 was indented with a tabulation.

**Review by Kerilk — CHANGES_REQUESTED** [review]
Thanks for the update. Another round of comments on those changes, and some I missed the first time around.

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: fixed.

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: fixed.

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: fixed

File: loader/icd.h
Comment (cdai2) [reply loader/icd.h]: fixed.

File: loader/icd_dispatch.c
Comment (cdai2) [reply loader/icd_dispatch.c]: fixed

File: loader/icd_dispatch_generated.c
Comment (cdai2) [reply loader/icd_dispatch_generated.c]: fixed

File: loader/icd.c
Comment (cdai2) [reply loader/icd.c]: I added on function named khrIcdInitializeTrace for it.

**Comment by cdai2** [comment]
Please review the fix again.  thanks @Kerilk

File: loader/windows/icd_windows_apppackage.cpp
Comment (Kerilk) [reply loader/windows/icd_windows_apppackage.cpp:25]: Leaving this open so @jenatali can take a look.

File: loader/icd_dispatch_generated.c
Comment (Kerilk) [inline loader/icd_dispatch_generated.c:24]: This file is generated, so it shouldn't be modified. Also I don't understand why this include would suddenly be required.
If the modification was required, https://github.com/KhronosGroup/OpenCL-ICD-Loader/blob/main/scripts/icd_dispatch_generated.c.mako would need to be updated before regenerating the dispatch.
```suggestion
#ifdef __cplusplus
```

**Review by Kerilk — CHANGES_REQUESTED** [review]
Thanks for the changes. The generated dispatch was only partially reverted, but aside from that looks good to me. @bashbaug and @jenatali can perhaps take a look and maybe spot what I missed.

File: loader/icd_dispatch_generated.c
Comment (cdai2) [reply loader/icd_dispatch_generated.c]: Fixed.  thanks

**Comment by cdai2** [comment]
@bashbaug  Please help to review.

File: README.md
Comment (bashbaug) [inline README.md:149]:
```suggestion
| OCL_ICD_ENABLE_TRACE             | Enable the trace mechanism          |  `export OCL_ICD_ENABLE_TRACE=True`<br/>`true, T, 1 can also be used here.`  |
```

File: loader/icd.h
Comment (bashbaug) [inline loader/icd.h:158]: This new if check should be moved inside of the one-trip do-while loop, rather than be placed outside.

For details, see: https://wiki.sei.cmu.edu/confluence/display/c/PRE10-C.+Wrap+multistatement+macros+in+a+do-while+loop

Applies here and the other places in this file where an if check was added.

**Review by bashbaug — CHANGES_REQUESTED** [review]
Mostly LGTM, just a few minor issues to fix.  This will be very useful!

I think we should also bump the Windows revision in OpenLC.rc after these changes?  This will make it easier to identify ICD loaders with the trace functionality vs. without.

Thanks!

**Review by bashbaug — COMMENTED** [review]
I am on vacation next week and will not be checking GitHub regularly.  I'm fine merging these changes as soon as my comments above are resolved.  Thank you!

**Comment by cdai2** [comment]
Thanks a lot for you to review.  I will update it immediately.

**Comment by cdai2** [comment]
> 

Which number should be bumped in OpenCL.rc?  REV or MAJOR or MINOR ? 

#define OPENCL_ICD_LOADER_VERSION_MAJOR 3
#define OPENCL_ICD_LOADER_VERSION_MINOR 0
#define OPENCL_ICD_LOADER_VERSION_REV   1

**Comment by cdai2** [comment]
@bashbaug   Please review. 
 I updated it according to your comments.  The OPENCL_ICD_LOADER_VERSION_REV  in OpenCL.rc is increased from 1 to 2.

**Review by Kerilk — APPROVED** [review]

File: README.md
Comment (cdai2) [reply README.md]: done

**Comment by cdai2** [comment]
Ping

**Comment by cdai2** [comment]
Can anyone help me to check or guide me to check the check error "Windows / msvc (v141, ON, Visual Studio 17 2022, x64, 90, 3.22.0) (pull_request) Cancelled after 2m — msvc (v141, ON, Visual Studio 17 2022, x64, 90, 3.22.0)"？   

I cannot reproduce any compilation error on my local machine with visual studio 2017.

**Comment by Kerilk** [comment]
I think you can safely ignore those, they are independent from your modifications. I'll let @bashbaug hit the merge button when he is satisfied that his requests have been incorporated.

File: loader/windows/icd_windows_apppackage.cpp
Comment (jenatali) [reply loader/windows/icd_windows_apppackage.cpp:25]: Weird. No idea why it'd do that.

Alternatively you could use:
```c++
#ifdef __cplusplus
extern "C" {
#endif

...

#ifdef __cplusplus
}
#endif
```

In the `icd.h` header to avoid having to re-declare it here with the extern.

**Review by jenatali — APPROVED** [review]
Looks good to me.

**Review by bashbaug — APPROVED** [review]
Latest changes LGTM - thanks!

**Comment by bashbaug** [comment]
There are three approvals for this PR now so I think we can safely merge this.

Note, the CI failures should be fixed by https://github.com/KhronosGroup/OpenCL-Headers/pull/198.

**Pull request merged**
